src/client/Client.cc

```diff
@@ -7999,9 +7999,21 @@ int Client::_readdir_cache_cb(dir_result_t *dirp, add_dirent_cb_t cb, void *p,
   while (true) {
     if (!dirp->inode->is_complete_and_ordered())
       return -EAGAIN;
+    
+    Dentry *dn = *pd;
+    int idx = pd - dir->readdir_cache.begin();
+    int r = _getattr(dn->inode, caps, dirp->perms);
+    if (r < 0)
+      return r;
+    
+    // the content of readdir_cache may change after _getattr(), so pd may be invalid iterator    
+    pd = dir->readdir_cache.begin() + idx;
+    if (pd >= dir->readdir_cache.end() && *pd != dn)
```

Comment:
sorry it should be
    if (pd >= dir->readdir_cache.end() || *pd != dn)

Reply:
maybe is “if (pd > dir->readdir_cache.end() || *pd != dn)”;
“if (pd == dir->readdir_cache.end())”  still take the break process.

Reply:
by the way, keep "if (pd == dir->readdir_cache.end())" in front of _getattr()